Prevent member name collisions in PowerShell classes

IISResetMe opened the pull request:
## PR Summary

Fix #8235, #7736

PowerShell currently allows definition of class with overlapping property and method names without warning, like so:

```powershell
class TestDuplicateMembers
{
  $A
  A(){}
}
```

Unlike the C# compiler (which throws a compilation error), we allow compilation of the above type definition, and ETS then implicitly hides the property behind the method at runtime - making the property implementation completely useless.

This change fixes that by making duplicate member name checks mutual between `DefineMethod()`/`DefineProperty()` - so that if a method with name `X` is already defined, an attempt to define property `X` will throw a `ParserError` (and vice versa) just like we already do for duplicate member names of the same member type

TODO:
 - [X] ~~Regression tests~~

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [X] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)


I'd consider this a bucket 3 breaking change (we'll start failing on defining types that are already broken but partially functional).

Change is obviously user-facing but I don't believe it warrants a documentation update since the previous documentation doesn't address the existing buggy behavior

IISResetMe committed:
Disallow class definitions with duplicate member names

Fix #8235

IISResetMe committed:
Move mutual check for duplicate class member names to PSType

As per @lzybkr's advice, duplicate member names should not be a parser concern since it's not a syntax violation.
This also avoids LINQ queries in the parser.

IISResetMe commented:
@SteveL-MSFT can you include this (or the linked issue) in the next committee review?

SteveL-MSFT commented:
@IISResetMe to set expectation, due to the holidays and unavailability of many committee members, this won't be reviewed until after the new year (and probably 2nd week of Jan at the earliest)

SteveL-MSFT commented:
@PowerShell/powershell-committee reviewed this, although .Net allows this, no language supports it and PowerShell should not either.  It is a technical breaking change although unlikely any customer is dependent on this since they would have defined a duplicate name member and not have used it.  We agree PowerShell should prevent this and have a parse time error.

IISResetMe committed:
Update PSType.cs

Fix CodeFactor issue

IISResetMe commented:
Thanks @SteveL-MSFT, will add regression tests for this then and we can move ahead! :)

IISResetMe committed:
Exclude ctor definitions from member name collision checks

The following class definition should still be valid (as in any other .NET language):

```
class A
{
  $A
  A(){}
}
```

IISResetMe committed:
Add negative parsing tests for member name collisions

IISResetMe committed:
Fix CodeFactor issue (SA1408)

iSazonov commented on src/System.Management.Automation/engine/parser/PSType.cs:
```diff
@@ -542,7 +542,7 @@ public void DefineMembers()
             private void DefineProperty(PropertyMemberAst propertyMemberAst)
             {
-                if (_definedProperties.ContainsKey(propertyMemberAst.Name))
+                if (_definedProperties.ContainsKey(propertyMemberAst.Name) || _definedMethods.ContainsKey(propertyMemberAst.Name) && !_definedMethods[propertyMemberAst.Name].Any(m => m.Item1.IsConstructor))
```
```suggestion
                if (_definedProperties.ContainsKey(propertyMemberAst.Name) || (_definedMethods.ContainsKey(propertyMemberAst.Name) && !_definedMethods[propertyMemberAst.Name].Any(m => m.Item1.IsConstructor)))
```

IISResetMe replied:
Already fixed 😅

iSazonov replied:
Could we avoid LINQ?

IISResetMe replied:
Actually we can avoid LINQ altogether - `IsConstructor` is name-bound, so if any `m` satisfies `IsConstructor == true`, they all will. We can simply inspect the first item in the list

iSazonov commented on src/System.Management.Automation/engine/parser/PSType.cs:
```diff
@@ -665,6 +665,15 @@ private PropertyBuilder EmitPropertyIl(PropertyMemberAst propertyMemberAst, Type
 
             private bool CheckForDuplicateOverload(FunctionMemberAst functionMemberAst, Type[] newParameters)
             {
+                if (_definedProperties.ContainsKey(functionMemberAst.Name) && !functionMemberAst.IsConstructor)
```
```suggestion
                if (!functionMemberAst.IsConstructor && _definedProperties.ContainsKey(functionMemberAst.Name))
```

iSazonov committed:
Exit early on property name collision checks for constructor definitions

Co-Authored-By: IISResetMe <IISResetMe@users.noreply.github.com>

IISResetMe committed:
Avoid unnecessary use of LINQ (.Any(this IEnumerable<T>)) in PSType

Since IsConstructor is already tied to the method name, we can safely assume that either all or none of them satisfy IsConstructor == true, so only checking the first list item is sufficient

IISResetMe committed:
[noop] re-run failed tests

iSazonov commented:
Yes, we are waiting.

IISResetMe committed:
Add comment regarding property name collision check logic

IISResetMe committed:
Expand previous commit

IISResetMe commented:
Can we turn off the "variable name looks like hungarian notation" rule flagged by CodeFactor? It's super noisy, and a cursory glance at S.M.A shows that we have over 900 parameter names alone that _look like_ hungarian notation to StyleCop, but I went through the list of name and only ~50 of them are actually named using hungarian notation:

```powershell
PS ~> $parameterNames = [psobject].Assembly.GetTypes().GetMethods().GetParameters().Name
PS ~> $looksHungarian = $parameterNames -cmatch '^[a-z]{2,3}[A-Z]+'
PS ~> $actuallyHungarian = $looksHungarian -notmatch '^(?:add|arg|ast|bad|do|eat|for|has|if|is|job|max|min|my|new|no|old|out|pre|ps|raw|run|set|sub|to|try|ui|use|yes)'
PS ~> $looksHungarian.Count
935
PS ~> $actuallyHungarian.Count
46
```

IISResetMe committed:
Trigger tests

iSazonov commented:
@IISResetMe We already have a PR to disable some CodeFactor rules.

IISResetMe commented on src/System.Management.Automation/engine/parser/PSType.cs:
```diff
@@ -542,7 +542,7 @@ public void DefineMembers()
             private void DefineProperty(PropertyMemberAst propertyMemberAst)
             {
-                if (_definedProperties.ContainsKey(propertyMemberAst.Name))
+                if (_definedProperties.ContainsKey(propertyMemberAst.Name) || (_definedMethods.ContainsKey(propertyMemberAst.Name) && !_definedMethods[propertyMemberAst.Name][0].Item1.IsConstructor))
```
Done :)

rjmholt replied:
Oh sorry, just meant a comment should explain what `_definedMethods[propertyMemberAst.Name][0].Item1` refers to, since it's a bit opaque. But the comment you added is also very useful

IISResetMe commented:
@iSazonov are we still waiting for the build pipeline to be fixed (CodeFactor issues)? Or do I need to address anything?

iSazonov commented:
@IISResetMe Fill free to ignore CodeFactor issues if they is not in your changed code. Maintainers will ask you to fix style issues before merge if they deem it necessary.

IISResetMe commented:
@daxian-dbw can you review this? :)

daxian-dbw commented:
@IISResetMe Sure, will do the review today.

daxian-dbw commented on src/System.Management.Automation/engine/parser/PSType.cs:
```diff
@@ -542,7 +542,15 @@ public void DefineMembers()
+                //       $C
+                //     }
+                //
+                if (_definedProperties.ContainsKey(propertyMemberAst.Name) || (_definedMethods.ContainsKey(propertyMemberAst.Name) && !_definedMethods[propertyMemberAst.Name][0].Item1.IsConstructor))
```
nit: Can you please define a local variable to hold `propertyMemberAst.Name` and replace `propertyMemberAst.Name` with that local variable in this method?
That will make this check less cumbersome and a bit more readable.

daxian-dbw commented:
@IISResetMe Your change doesn't fix the case where class A derives from class B:
```
class Bar { $c }
class Foo : Bar { c() {} }
$s = [Foo]::new()
$s.c()

Method invocation failed because [Foo] does not contain a method named 'c'.
At line:1 char:1
+ $s.c()
+ ~~~~~~
+ CategoryInfo          : InvalidOperation: (:) [], RuntimeException
+ FullyQualifiedErrorId : MethodNotFound
```
But to be honest, I don't know if it's worth the effort to fix that ...

vexx32 commented:
@IISResetMe are you gonna take that sitting down? :smile:

daxian-dbw commented:
@IISResetMe did work on something similar before -- flatten all implemented interfaces.
I know we need to fix the inheritance scenario for the sake of completeness, but personally, I just don't feel it's worth to do it ... :)

vexx32 commented:
Oh, that's right, he _did_. Should be a walk in the park, then. 😉 

But I get where you're coming from too, it _might_ not be worth it... but then again, it might end up being necessary -- after all, one of those members you're accidentally hiding might come from up the inheritance chain quite a ways, and it'll be very confusing if you hit _that_.

IISResetMe commented:
> Could this happen if you attempted to inherit from two disparate classes? Do PS classes even support that? 

No, and I can't think of a language that compiles against .NET that does... Multiple inheritance opens up a _whole new can of worms_ wrt member resolution ambiguity, so maybe let's not try and solve that as part of this PR ^__^

@daxian-dbw we can indeed leverage the same approach as with the interface flattening code for this. The extra lookups would be dependent on the user specifying a base class so it doesn't have to be super expensive for types that just inherit from Object. Will take a stab at it over the weekend

rjmholt commented:
Yeah, [the CLR doesn't support multiple inheritance, even for C++](https://stackoverflow.com/a/6258236)
